File: sorts/quick_sort/go/quicksort.go
package main

import (
	"fmt"
)

func main() {
	data := []int{5, 7, 3, 19, 4}
	//fmt.Println(partiton(data, 0, 0, len(data)-1))
	quicksort(data, 0, len(data)-1)
	fmt.Println(data)
}

func quicksort(data []int, first_index int, last_index int) {
	if first_index < last_index {

		pivot := partiton(data, first_index, first_index, last_index)
		quicksort(data, first_index, pivot-1)
		quicksort(data, pivot+1, last_index)
	}
}

func partiton(data []int, pivot_index int, first_index int, last_index int) int {
	left := first_index
	right := last_index
	pivot_value := data[pivot_index]
	go_while := true

	for go_while {
		for left <= right && pivot_value > data[left] {
			left = left + 1
		}

		for left <= right && pivot_value < data[right] {
			right = right - 1
		}

		if right <= left {
			break
		}

		tmp := data[left]
		data[left] = data[right]
		data[right] = tmp

	}

	return right

}

File: sorts/quick_sort/python/quicksort.py
def partition(data, pivot_index, first_index, last_index):
    pivot_value = data[pivot_index]
    left_index = first_index
    right_index = last_index


    while True:
        while left_index <= right_index and data[left_index] < pivot_value:
            left_index += 1

        while left_index <= right_index and data[right_index] > pivot_value:
            right_index -= 1

        if right_index <= left_index:
            break

        data[left_index], data[right_index] = data[right_index], data[left_index]

    return right_index

def quicksort(data, begin, end):
    if begin >= end:
        return

    pivot = partition(data, begin, begin, end)
    quicksort(data, begin, pivot - 1)
    quicksort(data, pivot + 1, end)

if __name__ == '__main__':
    data = [1, 5, 3, 7, 2, 9, 6]
    print quick(data, 0, len(data) - 1)
    print data
